service/matching/matching_engine.go

```diff
@@ -703,6 +711,20 @@ func (e *matchingEngineImpl) getHistoryForQueryTask(
 	return hist, resp.GetResponse().GetNextPageToken(), err
 }
 
+func (e *matchingEngineImpl) nonRetryableErrorsDropTask(task *internalTask, taskQueueName string, err error) {
+	e.logger.Info(err.Error(),
+		tag.WorkflowNamespaceID(task.event.Data.GetNamespaceId()),
+		tag.WorkflowID(task.event.Data.GetWorkflowId()),
+		tag.WorkflowRunID(task.event.Data.GetRunId()),
+		tag.WorkflowTaskQueueName(taskQueueName),
+		tag.TaskID(task.event.GetTaskId()),
+		tag.WorkflowEventID(task.event.Data.GetScheduledEventId()),
+		tag.Error(err),
+	)
+	// drop the task as otherwise task would be stuck in a retry-loop since the errors from within this helper is called are non-retryable
+	task.finish(nil)
```

Comment:
We need to emit a metric in this case and add an alert for it.

Reply:
I also think it would be beneficial for adding metrics grouped by each error type (be it internal, dataloss, etc.) This might make visibility better.
thoughts?

Reply:
Sure I don't have strong preference either way. This will only be triggered very rarely so we can also rely on logs to figure out details as long as the runbook is clear on what to search for. 

There's also a metrics.ServiceErrorTypeTag method you can use.

Reply:
sure thing - I currently have grouped them by the error types for better visibility.

Reply:
we're trying to reduce the number of namespace-scoped metrics.. maybe just a plain counter and rely on logs to go deeper?  imo, I think one metric is sufficient ("internal") and include the specific error type in the logs